### PR TITLE
Patterns

### PR DESCRIPTION
**Pull request by vantihovich**

### CODE REVIEW

File: patterns/behavioral/command/main.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+//is executed with the receiver
+type placeBetCommand struct {
+	aggregator *betAggregator
+}
+
+//the placeBet command itself, it implements the "command" interface with it`s "execute" method
+func (pb *placeBetCommand) execute() {
+	//increase the number of placed bets
+	pb.aggregator.totalBets += 1
+	fmt.Printf("The bet is placed, total number of bets is: %d\n", pb.aggregator.totalBets)
```
Comment (apaliavy) [inline patterns/behavioral/command/main.go:33]: I think it'd be better to return value here

File: patterns/behavioral/command/main.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+//adding method to betAggregator in order to create the instance of the command
```
Comment (apaliavy) [inline patterns/behavioral/command/main.go:36]: please start comment from the method name

File: patterns/behavioral/command/main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func (bm *BookMaker) executeCommand() {
+	for _, c := range bm.Commands {
+		c.execute()
+	}
+}
+
+func main() {
+
```
Comment (apaliavy) [inline patterns/behavioral/command/main.go:55]: space is not needed here

File: patterns/behavioral/command/main.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	tasks := []command{
+		ba.placeBet(),
+		ba.placeBet(),
+		ba.placeBet(),
+	}
+
+	placers := &BookMaker{}
+
+	for _, task := range tasks {
+		bookmaker := placers
```
Comment (apaliavy) [inline patterns/behavioral/command/main.go:67]: why?

File: patterns/behavioral/command/main.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+		bookmaker.Commands = append(bookmaker.Commands, task)
+	}
+
+	placers.executeCommand()
+
+	fmt.Println("total bets are:", ba.totalBets)
+
```
Comment (apaliavy) [inline patterns/behavioral/command/main.go:74]: space is not needed here

File: patterns/behavioral/observer/main.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
+package main
+
+import "fmt"
+
+type publisher interface {
+	register(Observer observer)   //adding new subscriber
```
Comment (apaliavy) [inline patterns/behavioral/observer/main.go:6]: addObserver()
removeObserver() 
... 

in this case comments are not needed

File: patterns/behavioral/observer/main.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	deRegister(Observer observer) //deleting subscriber
+	notify()                      //notifying subscriber
+}
+
+type observer interface { //observer interface
```
Comment (apaliavy) [inline patterns/behavioral/observer/main.go:11]: pointless comment

File: patterns/behavioral/observer/main.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+func (i item) updateAvailability() { // updating the state of object of interest
+	fmt.Printf("Item %q is now available\n", i.name)
+	i.in = true
+	i.notify() //calling notifying func for subscribed observers
+}
+
+func (i *item) register(o observer) { //adding new observer to the list
```
Comment (apaliavy) [inline patterns/behavioral/observer/main.go:34]: please try to avoid in-line comments

File: patterns/behavioral/observer/main.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+	for _, observer := range i.observerList {
+		observer.update(i.name)
+	}
+}
+
+func removeFromSlice(observerList []observer, observerToRemove observer) []observer { // makes the list shorter by deleting the observer with signed name
+	observerListLength := len(observerList)
+	for i, observer := range observerList {
+		if observerToRemove.getID() == observer.getID() {
+			observerList[observerListLength-1], observerList[i] = observerList[i], observerList[observerListLength-1]
```
Comment (apaliavy) [inline patterns/behavioral/observer/main.go:52]: hmmm maybe it'd be better to store observers in map? (not sure) 
in that case the process of observer removal might looks much simpler

File: patterns/behavioral/observer/main.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+			return observerList[:observerListLength-1]
+		}
+	}
+	return observerList
+}
+
+type customer struct { //the object that is interested in notification
```
Comment (apaliavy) [inline patterns/behavioral/observer/main.go:59]: based on the comment is seems like `subscriber` name is the best match

File: patterns/creational/builder/main.go [hunk 11]
```diff
@@ -0,0 +1,147 @@
+package main
+
+import "fmt"
+
+type pizza struct {
```
Comment (apaliavy) [inline patterns/creational/builder/main.go:5]: 👨‍🍳

File: patterns/creational/builder/main.go [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+func getPizzaMaker(pizzaType string) pizzaBuilder {
+	if pizzaType == "pepperoni" {
+		return &pepperoniBuilder{}
+	}
+
+	if pizzaType == "vesuvius" {
+		return &vesuviusBuilder{}
+	}
+
+	return nil
```
Comment (apaliavy) [inline patterns/creational/builder/main.go:29]: i'd say we should return an error here

File: patterns/structural/proxy/main.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	User string
+}
+
+type ProxyObject struct {
+	Object
+	Reader
+}
+
+func (po *ProxyObject) Read() {
+	if po.Reader.User == "Alice" {
```
Comment (apaliavy) [inline patterns/structural/proxy/main.go:30]: seems like nil check is missing

File: patterns/behavioral/command/main.go (hunk 1 above)
Comment (vantihovich) [reply patterns/behavioral/command/main.go:33]: done

File: patterns/behavioral/command/main.go (hunk 2 above)
Comment (vantihovich) [reply patterns/behavioral/command/main.go]: done

File: patterns/behavioral/command/main.go (hunk 3 above)
Comment (vantihovich) [reply patterns/behavioral/command/main.go]: done

File: patterns/behavioral/command/main.go (hunk 4 above)
Comment (vantihovich) [reply patterns/behavioral/command/main.go]: removed unnecessary variable, but without this variable the warning appears: "should replace loop with placers.Commands = append(placers.Commands, tasks...) (S1011) go-staticcheck", however the code is compiled and executed, so leaving it with warning.

File: patterns/behavioral/command/main.go (hunk 5 above)
Comment (vantihovich) [reply patterns/behavioral/command/main.go]: done

File: patterns/behavioral/observer/main.go (hunk 6 above)
Comment (vantihovich) [reply patterns/behavioral/observer/main.go]: done

File: patterns/behavioral/observer/main.go (hunk 7 above)
Comment (vantihovich) [reply patterns/behavioral/observer/main.go]: removed

File: patterns/behavioral/observer/main.go (hunk 8 above)
Comment (vantihovich) [reply patterns/behavioral/observer/main.go]: done

File: patterns/behavioral/observer/main.go (hunk 9 above)
Comment (vantihovich) [reply patterns/behavioral/observer/main.go:59]: tried to store observers in map , but that caused additional "issues" to solve like: storing  the observer interface value as a key value for map(to have an ability to link event and notifier via interface), converting string value(subscribers name) to observer interface type to pass it to map, not solved, but looks like causes code complication in part of passing the values, so left as is with slice.

File: patterns/behavioral/observer/main.go (hunk 10 above)
Comment (vantihovich) [reply patterns/behavioral/observer/main.go]: done

File: patterns/creational/builder/main.go (hunk 11 above)
Comment (vantihovich) [reply patterns/creational/builder/main.go:8]: yep))

File: patterns/creational/builder/main.go (hunk 12 above)
Comment (vantihovich) [reply patterns/creational/builder/main.go]: done

File: patterns/structural/proxy/main.go (hunk 13 above)
Comment (vantihovich) [reply patterns/structural/proxy/main.go:33]: added

File: patterns/structural/proxy/main.go [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+	Object
+	Reader
+}
+
+func (po *ProxyObject) Read() {
+	err := errors.New("username not provided")
+
+	if po.Reader.User == "Alice" {
+		fmt.Println("User has an access")
+		fmt.Printf("Object:%s, Reader:%s", po.Object, po.Reader)
```
Comment (iamnotjustice) [inline patterns/structural/proxy/main.go:35]: Proxy should not change the interface nor change the behavior. So proxy here is adding some additional ("operational") things like access control, this is great. And the interface is intact, this is also great. But one thing is missing here - calling the original wrapped object's method inside this proxy.

**Review by iamnotjustice — APPROVED** [review]

File: patterns/behavioral/command/main.go (hunk 4 above)
Comment (apaliavy) [reply patterns/behavioral/command/main.go]: Why didn't you apply the proposed change?

Comment (vantihovich) [reply patterns/behavioral/command/main.go]: Fixed in https://github.com/vantihovich/go_tasks/pull/15/commits/6f11cbb7e889f3c355d6665d53cf5fa1228f3db5

**Pull request merged**